### PR TITLE
[UNR-927] Better error reporting during schema generation

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Fail schema generation if there are replicated/handover properties, RPCs, or subobjects that contain only non-alphanumeric characters, or leading digits, since they will produce invalid schema names.
Also, report all the errors that are found across all supported classes rather than early outing on the first error.

#### Release note
Bugfix: Report invalid name errors during schema generation instead of when launching a deployment.

#### Tests
Create replicated/handover properties, RPCs, or subobjects with invalid names:
- names only consisting of non-alphanumeric characters
- leading digits after removing non-alphanumeric characters
- clashing identifier names after removing non-alphanumeric characters

image
image

#### Documentation
Need to figure out where and how to document our additional restrictions on variable names.
@NotRobSheridan 

#### Primary reviewers
@Vatyx @joshuahuburn

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp [hunk 1]
```diff
@@ -541,47 +541,31 @@ void GenerateSubobjectSchemaForActor(FComponentIdGenerator& IdGenerator, UClass*
...
 	GenerateActorIncludes(Writer, TypeInfo);
 
+	FSubobjectMap Subobjects = GetAllSubobjects(TypeInfo);
+
 	bool bHasComponents = false;
-	TSet<UObject*> SeenComponents;
-	int32 CurrentOffset = 1;
 
-	for (auto& PropertyPair : TypeInfo->Properties)
+	for (auto& It : Subobjects)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp]: Could this be `const auto&`?

Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp]:
```suggestion
	for (const auto& It : Subobjects)
```

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.cpp [hunk 2]
```diff
@@ -53,7 +53,10 @@ FString AlphanumericSanitization(const FString& InString)
 FString UnrealNameToSchemaComponentName(const FString& UnrealName)
 {
 	FString SchemaTypeName = UnrealNameToSchemaName(UnrealName);
-	SchemaTypeName[0] = FChar::ToUpper(SchemaTypeName[0]);
+	if (!SchemaTypeName.IsEmpty())
+	{
+		SchemaTypeName[0] = FChar::ToUpper(SchemaTypeName[0]);
+	}
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.cpp]: If this is empty would we want to error?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/Utils/DataTypeUtilities.cpp]: We will error in `CheckIdentifierNameValidity` (which calls this).

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 3]
```diff
@@ -63,7 +63,24 @@ int GenerateCompleteSchemaFromClass(FString SchemaPath, int ComponentId, TShared
 	return NumComponents;
 }
 
-bool CheckIdentifierNameValidity(TSharedPtr<FUnrealType> TypeInfo)
+bool CheckSchemaName(FString Name, FString Identifier, FString Category)
```
Comment (ieva-improbable) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Rename to `CheckSchemaNameValidity`?

**Pull request merged**